### PR TITLE
validate_json_schema() migrated

### PR DESCRIPTION
**Pull request by SarahJohnsonONS**
`validate_json_schema()` migrated from GSSCogs repo. Tests all passing. README.md added.

Also set up pyproject.toml and poetry.lock. Run `poetry install` to install dependencies.

### CODE REVIEW

**Review by NickPapONS — DISMISSED** [review]
Migration looks correct, all tests pass. Approved!

edit: Never mind. I can approve but I still don't have write access to ONS Digital (even though I just accepted an invite to the github a few minutes ago). You might have to get someone else to have a look, sorry.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+[tool.poetry]
+name = "dpytools"
+version = "0.1.0"
+description = "Simple reusable python resources for digital publishing"
+authors = ["Your Name <you@example.com>"]
+license = "MIT"
```
Comment (mikeAdamss) [inline pyproject.toml:6]: can you add this please. Its just a file that sits in the repo base.

use this: https://github.com/GSS-Cogs/dp-python-tools/blob/main/LICENSE and change the date to 2024 please.

Comment (mikeAdamss) [reply pyproject.toml:6]: can you add a licence section to the readme as well please, use the one from here as an example: https://github.com/ONSdigital/dp , again just change the year.

its a faff but we need to be more formal with our boiler plate now.

File: dpytools/validation/json/validation.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    error_msg: Optional[str] = None,
+    indent: Optional[int] = None,
+):
+    """
+    Validate a JSON file against a schema.
+
+    Either `data_dict` or `data_path` must be provided.
+
+    `error_msg` and `indent` can be used to format the error message if validation fails.
+    """
```
Comment (mikeAdamss) [inline dpytools/validation/json/validation.py:23]: a general point but you can restructure this to flow a bit better, at the moment its (a) dealing with data path/dict (b) dealing with schema (c) dealing with data path/dict again.

I'd probably deal with the schema stuff first then the rest can just be data dict/path logic then into the validation.

I'd also have a think about assertions, very very useful thing for quickly and cleanly sanitising inputs, eg:

```python
assert not all(data_dict, data_path), "dont provide both dummy"
``` 

which is effectively:

```python
if all([data_path, data_dict)]:
      raise AssertionError("dont provide both dummy")
```

handy thing and maps quite nicely to what you're doing here, you just have to reverse your thinking a bit.

File: dpytools/validation/json/validation.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        # Check `data_path` exists
+        if not data_path.exists():
+            raise ValueError(f"Data path '{data_path}' does not exist")
+        with open(data_path, "r") as f:
+            data_to_validate = json.load(f)
+
+    # Validate data against schema
+    try:
+        jsonschema.validate(data_to_validate, schema_from_path)
+    # TODO Handle jsonschema.SchemaError?
```
Comment (mikeAdamss) [inline dpytools/validation/json/validation.py]: we. an probably lose the TODO, if you mean what happens if the json schema package itself errors, not our problem.

File: tests/test_json_validation.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
+from pathlib import Path
+from jsonschema import ValidationError
+
+import pytest
+from dpytools.validation.json.validation import validate_json_schema
+
+
+def test_validate_json_schema_data_path():
```
Comment (mikeAdamss) [inline tests/test_json_validation.py:8]: you dont need to assert none. Just run the function, if. it raises the test will fail with that exception - thats fine.

File: tests/test_json_validation.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    pipeline_config_schema = "tests/test_cases/pipeline_config_schema.json"
+    pipeline_config = {
+        "schema": "airflow.schemas.ingress.sdmx.v1.schema.json",
+        "required_files": [{"matches": "*.sdmx", "count": 1}],
+        "supplementary_distributions": [{"matches": "*.sdmx", "count": 1}],
+        "priority": 1,
+        "contact": ["jobloggs@ons.gov.uk"],
+        "pipeline": "default",
+    }
+    assert (
```
Comment (mikeAdamss) [inline tests/test_json_validation.py]: same as above

**Review by mikeAdamss — COMMENTED** [review]
Added some minor comments (my inner pedant is strong today). Code works and the readme is excellent though 👍

File: dpytools/validation/json/validation.py (hunk 3 above)
Comment (SarahJohnsonONS) [reply dpytools/validation/json/validation.py]: There are 2 main jsonschema errors, a ValidationError and a SchemaError, for if there's a problem with the schema itself - I wasn't sure if we wanted to handle those as well.

Comment (mikeAdamss) [reply dpytools/validation/json/validation.py]: oh, a malformed schema? interesting, didn't think of that - I see your point.

but no, it'll get raised as an uncaught error which is what we want. We'll need to catch it where we call the validation function (as it'd indicate a mess up by us not the data submitter, no point notifying them).

File: LICENSE [hunk 6]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2023 ONS Digital
```
Comment (mikeAdamss) [inline LICENSE]: 2024

**Review by mikeAdamss — COMMENTED** [review]
good job and good question. tweak that year and I shall approve 😄

**Review by mikeAdamss — APPROVED** [review]
LGTM

**Pull request merged**